android/src/main/java/com/squareup/sdk/reader/react/StoreCustomerCardModule.java

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2018 Square Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package com.squareup.sdk.reader.react;
+
+import android.app.Activity;
+import android.os.Handler;
+import android.os.Looper;
+import com.facebook.react.bridge.Promise;
+import com.facebook.react.bridge.ReactApplicationContext;
+import com.facebook.react.bridge.ReactContextBaseJavaModule;
+import com.facebook.react.bridge.ReactMethod;
+import com.squareup.sdk.reader.ReaderSdk;
+import com.squareup.sdk.reader.checkout.Card;
+import com.squareup.sdk.reader.core.CallbackReference;
+import com.squareup.sdk.reader.core.Result;
+import com.squareup.sdk.reader.core.ResultError;
+import com.squareup.sdk.reader.crm.StoreCardActivityCallback;
+import com.squareup.sdk.reader.crm.StoreCustomerCardErrorCode;
+import com.squareup.sdk.reader.react.internal.ErrorHandlerUtils;
+import com.squareup.sdk.reader.react.internal.ReaderSdkException;
+import com.squareup.sdk.reader.react.internal.converter.CardConverter;
+
+class StoreCustomerCardModule extends ReactContextBaseJavaModule {
+    // Define all the store customer card debug codes and messages below
+    // These error codes and messages **MUST** align with iOS error codes and javascript error codes
+    // Search KEEP_IN_SYNC_STORE_CUSTOMER_CARD_ERROR to update all places
+
+    // react native module debug error codes
+    private static final String RN_STORE_CUSTOMER_CARD_ALREADY_IN_PROGRESS = "rn_store_customer_card_already_in_progress";
+
+    // react native module debug messages
+    private static final String RN_MESSAGE_STORE_CUSTOMER_CARD_ALREADY_IN_PROGRESS = "A store customer card operation is already in progress. Ensure that the in-progress store customer card is completed before calling startStoreCardAsync again.";
+
+    private volatile CallbackReference storeCardCallbackRef;
+    private final Handler mainLooperHandler;
+
+    public StoreCustomerCardModule(ReactApplicationContext reactContext) {
+        super(reactContext);
+        mainLooperHandler = new Handler(Looper.getMainLooper());
+    }
+
+    @Override
+    public String getName() {
+        return "RNReaderSDKStoreCustomerCard";
+    }
+
+    @ReactMethod
+    public void startStoreCard(final String customerId, final Promise promise) {
+        if (storeCardCallbackRef != null) {
```

Comment:
This works... but we already check in the Reader SDK layer and fail if there's already an operation in flight, so this feels redundant.  And it means there are two potential error results for storing cards, one here and one in an error result to the callback.

Reply:
We take this same approach in both iOS and Android to enforce the same behavior and error code on javascript layer. in addition, this is a protection for not leaking storeCardCallbackRef.